File: casper/src/rust/helper/rho_logger_contract.rs
// See casper/src/test/scala/coop/rchain/casper/helper/RhoLoggerContract.scala

use models::rhoapi::{ListParWithRandom, Par};
use rholang::rust::interpreter::{
    contract_call::ContractCall,
    errors::{illegal_argument_error, InterpreterError},
    pretty_printer::PrettyPrinter,
    rho_type::RhoString,
    system_processes::ProcessContext,
};

pub struct RhoLoggerContract;

impl RhoLoggerContract {
    pub async fn std_log(
        ctx: ProcessContext,
        message: (Vec<ListParWithRandom>, bool, Vec<Par>),
    ) -> Result<Vec<Par>, InterpreterError> {
        let mut pretty_printer = PrettyPrinter::new();

        let is_contract_call = ContractCall {
            space: ctx.space.clone(),
            dispatcher: ctx.dispatcher.clone(),
        };

        if let Some((_, _, _, args)) = is_contract_call.unapply(message) {
            match args.as_slice() {
                [log_level_par, par] => {
                    if let Some(log_level) = RhoString::unapply(log_level_par) {
                        let msg = pretty_printer.build_string_from_message(par);

                        match log_level.as_str() {
                            "trace" => {
                                println!("trace: {}", msg);
                                Ok(vec![])
                            }
                            "debug" => {
                                println!("debug: {}", msg);
                                Ok(vec![])
                            }
                            "info" => {
                                println!("info: {}", msg);
                                Ok(vec![])
                            }
                            "warn" => {
                                println!("warn: {}", msg);
                                Ok(vec![])
                            }
                            "error" => {
                                println!("error: {}", msg);
                                Ok(vec![])
                            }
                            _ => Err(illegal_argument_error("std_log")),
                        }
                    } else {
                        Err(illegal_argument_error("std_log"))
                    }
                }
                _ => Err(illegal_argument_error("std_log")),
            }
        } else {
            Err(illegal_argument_error("std_log"))
        }
    }
}

File: casper/src/test/scala/coop/rchain/casper/helper/RhoLoggerContract.scala
package coop.rchain.casper.helper
import cats.Functor
import cats.syntax.functor._
import cats.effect.Concurrent
import coop.rchain.metrics.Span
import coop.rchain.models.{ListParWithRandom, Par}
import coop.rchain.rholang.interpreter.SystemProcesses.ProcessContext
import coop.rchain.rholang.interpreter.{ContractCall, PrettyPrinter, RhoType}
import coop.rchain.shared.{Log, LogSource}

object RhoLoggerContract {
  val prettyPrinter = PrettyPrinter()

  //TODO extract a `RhoPatterns[F]` algebra that will move passing the Span, the Dispatcher, and the Space parameters closer to the edge of the world
  def handleMessage[F[_]: Log: Functor: Concurrent: Span](
      ctx: ProcessContext[F]
  )(message: Seq[ListParWithRandom], isReplay: Boolean, previousOutput: Seq[Par]): F[Seq[Par]] = {
    val isContractCall = new ContractCall(ctx.space, ctx.dispatcher)

    (message, isReplay, previousOutput) match {
      case isContractCall(_, _, _, Seq(RhoType.String(logLevel), par)) =>
        val msg         = prettyPrinter.buildString(par)
        implicit val ev = LogSource.matLogSource

        logLevel match {
          case "trace" => Log[F].trace(msg).map(_ => Seq.empty[Par])
          case "debug" => Log[F].debug(msg).map(_ => Seq.empty[Par])
          case "info"  => Log[F].info(msg).map(_ => Seq.empty[Par])
          case "warn"  => Log[F].warn(msg).map(_ => Seq.empty[Par])
          case "error" => Log[F].error(msg).map(_ => Seq.empty[Par])
        }
    }
  }
}
